crates/matrix-sdk/src/event_cache/linked_chunk/as_vector.rs

```diff
@@ -0,0 +1,529 @@
+
+                            self.chunks.insert(next_chunk_index, (*new, 0));
+                        }
+
+                        (None, None) => {
+                            unreachable!(
+                                "Inserting new chunk with no previous nor next chunk identifiers \
+                                is impossible"
+                            );
+                        }
+                    }
+                }
+
+                Update::RemoveChunk(expected_chunk_identifier) => {
+                    let chunk_index = self
+                        .chunks
+                        .iter()
+                        .position(|(chunk_identifier, _)| {
+                            chunk_identifier == expected_chunk_identifier
+                        })
+                        // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and assuming
+                        // `Self::chunks` is correctly initialized, it is not possible to remove a
+                        // chunk that does not exist. If this predicate fails, it means
+                        // `LinkedChunk` or `Updates` contain a bug.
+                        .expect("Removing a chunk: The chunk is not found");
+
+                    // It's OK to ignore the result. The `chunk_index` exists because it's been
+                    // found, and we don't care about its associated value.
+                    let _ = self.chunks.remove(chunk_index);
+                }
+
+                Update::PushItems { position_hint: position, items } => {
+                    let expected_chunk_identifier = position.chunk_identifier();
+
+                    let (chunk_index, offset, chunk_length) = {
+                        let control_flow = self.chunks.iter_mut().enumerate().try_fold(
+                            position.index(),
+                            |offset, (chunk_index, (chunk_identifier, chunk_length))| {
+                                if chunk_identifier == &expected_chunk_identifier {
+                                    ControlFlow::Break((chunk_index, offset, chunk_length))
+                                } else {
+                                    ControlFlow::Continue(offset + *chunk_length)
+                                }
+                            },
+                        );
+
+                        match control_flow {
+                            ControlFlow::Break(value) => Some(value),
+                            ControlFlow::Continue(..) => None,
+                        }
+                    }
+                    // SAFETY: Assuming `LinkedChunk` and `Updates` are not buggy, and assuming
+                    // `Self::chunks` is correctly initialized, it is not possible to push items on
+                    // a chunk that does not exist. If this predicate fails, it means `LinkedChunk`
+                    // or `Updates` contain a bug.
+                    .expect("Pushing items: The chunk is not found");
```

Comment:
Could this be replaced with a loop? I find it hard to follow:

- we introduce an option at the end, that we try to unwrap immediately thereafter
- ending with matching that control_flow is `Continue(..)` means that we never found the item, which could be express much more explicitly (e.g. by keeping `let mut found = false` flag set to `true` when found, with an explicit `break` statement in the loop)

Reply:
I've removed the `Option`, indeed it's pointless. I wish [`ControlFlow::break_value`](https://doc.rust-lang.org/std/ops/enum.ControlFlow.html#method.break_value) lands soon in stable.

I like the `try_fold` approach better than a `for` loop though.